### DISCUSSION TITLE
Visualising the last layer feature map with tSNE

### DISCUSSION THREAD

**Discussion by akashtadwai** [ContinualAI/avalanche#509] [Q&A] [answered]
Hi,
I am using a SimpleMLP model with num_classes=15, hidden_size=100 with a GEM plugin. I wanted to visualize the last feature map before the classifier head with tSNE so that I could visually see the 2D projection of what my model is learning (similar to what is done in [this](https://learnopencv.com/t-sne-for-feature-visualization/))  in subsequent tasks. For that, I should have access to the test_stream labels and the model after the test_stream. Can anyone help me with this?  My basic code for the model is uploaded [here](https://github.com/ReethuVinta/AnomalyDetection/blob/master/avalanche/avalanche_ids.py).

**Comment by vlomonaco** [comment, marked as answer]
Hi @akashtadwai! Adding a metric to visualize feature maps is something we definitely would like to integrate in the near future. I suggest to go through the [From Zero to Hero tutorial](https://avalanche.continualai.org/from-zero-to-hero-tutorial/01_introduction) to see how you can add custom metrics. 

As for your request, you can access the dataset of each experience at any time with `test_stream[idx_exp].dataset` and you can extract targets iterating trough it with a data loader or by accessing the targets field: `test_stream[idx_exp].dataset.targets` directly.

The final model at the end of the training can be recovered with `model = cl_strategy.model`.

Let me know if you need further assistance! :)

**Reply by akashtadwai** [reply]
Thanks for the reply @vlomonaco . What I exactly need is,
I have defined ```model = SimpleMLP(num_classes=15,input_size=70,hidden_size=100)``` and  ```cl_strategy = GEM(
    model, optimizer = Adam(model.parameters()),patterns_per_exp=4400,
    criterion = CrossEntropyLoss(), train_mb_size=128, train_epochs=5, eval_mb_size=128,
    evaluator=eval_plugin,device=device)```  
I need to redefine the following params according to my need.
```python
for experience in generic_scenario.train_stream:
    res = cl_strategy.train(experience)
    print('Training completed')
    features  = []
    labels = []
    current_test_set = generic_scenario.test_stream
    model.eval() # what should I replace this with??
    for idx,data in enumerate(current_test_set):
         # is this correct? 
         # # If isn't  what shouldI replace this with to get the output of last layer before classification layer
        output = cl_strategy.model.forward(cur_test_set[experience.current_experience,idx].dataset)
        labels+=cur_test_set[experience.current_experience,idx].dataset.targets
        current_outputs = output.cpu().numpy()
        if features is not None:
            features = np.concatenate((features, current_outputs))
        else:
            features = current_features
       # some other code to Visualise features...
```

**Reply by AndreaCossu** [reply]
Hi @akashtadwai ! If your model does not output the activations you need, I think you have to rely on forward hooks by Pytorch to get them and then plot them. Avalanche does not manage the internals of your model.
Instead, if your model already outputs the activations you want, you can plot them by calling `out = model(data)` directly in your main, for example after each `cl_strategy.train(experience)` call. You can then apply t-SNE on `out`.

If you want this process at a fine-grained level (e.g. after each iteration) or if you want it integrated with `TensorboardLogger` and `WandBLogger`  you have to write a Plugin Metric which overwrite the appropriate callback (e.g. `after_training_iteration`, `after_forward`...). You can access all the properties of the strategy within the Plugin Metric itself. Check out how to implement a Plugin Metric [here](https://avalanche.continualai.org/from-zero-to-hero-tutorial/05_evaluation#implement-your-own-metric).

**Reply by EdenBelouadah** [reply]
If I understand your two questions: 
1- You can access to the trained model by using cl_strategy.model after the training of your model 
2- To pass to the evaluation mode, you use cl_strategy.model.eval() 
3- You access to the dataset used in the benchmark and manually wrap it in a loader to extract features of the data that you need.. Here is a pseudo-code: 

```
for experience in generic_scenario.train_stream:
    cl_strategy.train(experience)
    current_test_set = generic_scenario.test_stream[experience.current_experience].dataset    
    loader = torch.utils.data.DataLoader(current_test_set, ...)
    cl_strategy.model.eval() 
    for data in loader:
           # extract features
           outputs = cl_strategy.model(data) # you need to pass it first by Variable
    cl_strategy.model.train() 

```

I hope my solution is not dirty compared to the way Avalanche is built.

**Reply by akashtadwai** [reply]
Thanks for your replies @AndreaCossu  @EdenBelouadah !
> cl_strategy.model.train()  

@EdenBelouadah  Can you describe why do we need to use the train again in the last line??!  And isn't cl_strategy.model() and model (SimpleMLP) defined at first same. We are updating the SimpleMLP model with custom plugins like GEM, isn't it??  

And I will try to save the model after each task by registering hooks and get the activations of the hidden layer after the whole training and testing process in a separate script so that I could easily write model.eval() and generate tSNE plots for test_stream. I think I should be able to do that.

**Reply by EdenBelouadah** [reply]
@akashtadwai, I just wanted to remind you to return to the training mode in case you will continue training your model after extracting the scores (I think it's even done automatically in the BaseStrategy's training loop). If you don't need it, remove it. 

Yes, SimpleMLP and strategy.model are the same. It's just that strategy.model points to the trained model. However the instantiation of SimpleMLP is randomly initialized. 

@AndreaCossu, please correct me if I'm wrong.

**Reply by vlomonaco** [reply]
@akashtadwai let us know if we can mark your question as "answered", thanks! :)

**Reply by akashtadwai** [reply]
@vlomonaco  Yes Thanks! I have done it using hooks!

**Reply by akashtadwai** [reply]
> test_stream[idx_exp].dataset

 @vlomonaco @AndreaCossu  Can someone clarify what exactly the ```train_stream[exp_idx].dataset[idx]``` returns.   
I was doing stg like this,  
``` python
for idx in range(len(train_stream)):
    print(type(train_stream[idx].dataset[0]))
    break
```
I was getting at tuple with 3 attributes when I ran locally but when I saw the original code in  [avalanche_dataset.py](https://github.com/ContinualAI/avalanche/blob/68290d32f9710639d5fda6d17da4072a4458f702/avalanche/benchmarks/utils/avalanche_dataset.py#L672) It says that it consists of a tuple with 4 attributes.  
Can you exactly specify what are the attributes in the tuple?    
I guess the attributes I got are (features[idx],label,task_label). Is this correct??? and what does ```*element[2:]``` mean in the code  [avalanche_dataset.py](https://github.com/ContinualAI/avalanche/blob/68290d32f9710639d5fda6d17da4072a4458f702/avalanche/benchmarks/utils/avalanche_dataset.py#L672).

**Reply by vlomonaco** [reply]
Hi @akashtadwai! As you correctly derived `train_stream[exp_idx].dataset[idx]` returns an AvalancheDataset. an [AvalancheDataset](https://avalanche-api.continualai.org/autoapi/avalanche/benchmarks/utils/avalanche_dataset/?highlight=avalanchedataset#avalanche.benchmarks.utils.avalanche_dataset.AvalancheDataset) returns a tuple that may contain a variable number of elements based on the specific dataset/task. For classification  (data, label, task_label) is often enough. However, for detection or segmentation you often need more data: (data, label, bounding_box, task_label) for example. So in general the AvalancheDataset returns a tuple of this form (data, label, ... , task_label).

**Comment by AndreaCossu** [comment]
You are correct @EdenBelouadah, if you have direct access to `model` you can use that instance. Otherwise, if you have access to strategy you can access the model by `strategy.model`. This is the same.
